src/cljs/rems/actions.cljs

```diff
@@ -12,61 +14,101 @@
+   {:db (assoc db
+               ::todo-applications-query query
+               ::fetching-todo-applications? true)}))
 
 (rf/reg-event-db
  ::fetch-todo-applications-result
- (fn [db [_ result]]
+ (fn [db [_ result query]]
+   ;; fetch again if the query that just finished was not the latest
+   (when-not (= query (::todo-applications-query db))
+     (rf/dispatch [::fetch-todo-applications (::todo-applications-query db)]))
    (-> db
        (assoc ::todo-applications result)
-       (dissoc ::loading-todo-applications?))))
+       (dissoc ::initializing-todo-applications?
+               ::fetching-todo-applications?))))
 
 (rf/reg-sub
  ::todo-applications
  (fn [db _]
    (::todo-applications db)))
 
 (rf/reg-sub
- ::loading-todo-applications?
+ ::initializing-todo-applications?
+ (fn [db _]
+   (::initializing-todo-applications? db)))
+
+(rf/reg-sub
+ ::searching-todo-applications?
  (fn [db _]
-   (::loading-todo-applications? db)))
+   (and (::fetching-todo-applications? db)
+        (not (::initializing-todo-applications? db)))))
 
 ;;;; handled applications
 
 (rf/reg-event-fx
  ::fetch-handled-applications
- (fn [{:keys [db]} _]
-   (fetch "/api/applications/handled"
-          {:handler #(rf/dispatch [::fetch-handled-applications-result %])})
-   {:db (assoc db ::loading-handled-applications? true)}))
+ (fn [{:keys [db]} [_ query]]
+   ;; do only one fetch at a time - will retry after the pending fetch is finished
+   (when-not (::fetching-handled-applications? db)
+     (fetch "/api/applications/handled"
+            {:url-params (when query
+                           {:query query})
+             :handler #(rf/dispatch [::fetch-handled-applications-result % query])
+             ;; TODO: show error message
+             :error-handler #(rf/dispatch [::fetch-handled-applications-result nil query])}))
+   {:db (assoc db
+               ::handled-applications-query query
+               ::fetching-handled-applications? true)}))
 
 (rf/reg-event-db
  ::fetch-handled-applications-result
- (fn [db [_ result]]
+ (fn [db [_ result query]]
+   ;; fetch again if the query that just finished was not the latest
+   (when-not (= query (::handled-applications-query db))
+     (rf/dispatch [::fetch-handled-applications (::handled-applications-query db)]))
    (-> db
        (assoc ::handled-applications result)
-       (dissoc ::loading-handled-applications?))))
+       (dissoc ::initializing-handled-applications?
+               ::fetching-handled-applications?))))
 
 (rf/reg-sub
  ::handled-applications
  (fn [db _]
    (::handled-applications db)))
 
 (rf/reg-sub
- ::loading-handled-applications?
+ ::initializing-handled-applications?
+ (fn [db _]
+   (::initializing-handled-applications? db)))
+
+(rf/reg-sub
+ ::searching-handled-applications?
  (fn [db _]
-   (::loading-handled-applications? db)))
+   (and (::fetching-handled-applications? db)
+        (not (::initializing-handled-applications? db)))))
```

Comment:
the -handled- and -todo- are a bit repetitive, but I guess two copies is fine

Reply:
I'll dedupe them later when adding the search to other pages.